messaging/src/main/java/org/axonframework/messaging/deadletter/InMemoryDeadLetterQueue.java

```diff
@@ -0,0 +1,373 @@
+
+        return letter;
+    }
+
+    private DeadLetter<T> buildLetter(QueueIdentifier identifier, T deadLetter, Throwable cause) {
+        Instant deadLettered = clock.instant();
+        Instant expiresAt = cause == null ? deadLettered : deadLettered.plus(expireThreshold);
+        return new GenericDeadLetter<>(identifier,
+                                       deadLetter,
+                                       cause,
+                                       deadLettered,
+                                       expiresAt,
+                                       this::acknowledge,
+                                       this::requeue);
+    }
+
+    private void acknowledge(DeadLetter<T> letter) {
+        QueueIdentifier queueId = letter.queueIdentifier();
+        Deque<DeadLetter<T>> sequence = deadLetters.get(queueId);
+        sequence.remove(letter);
+        logger.trace("Removed letter [{}].", letter.message().getIdentifier());
+        if (sequence.isEmpty()) {
+            logger.trace("Queue [{}] is empty and will be removed.", queueId);
+            deadLetters.remove(queueId);
+        }
+        takenSequences.remove(queueId);
+    }
+
+    private void requeue(DeadLetter<T> letter) {
+        QueueIdentifier queueId = letter.queueIdentifier();
+        Deque<DeadLetter<T>> sequence = deadLetters.get(queueId);
+        sequence.remove(letter);
+
+        Instant newExpiresAt = clock.instant().plus(expireThreshold);
+        DeadLetter<T> updatedLetter =
+                new GenericDeadLetter<>(letter, newExpiresAt, this::acknowledge, this::requeue);
+        sequence.addFirst(updatedLetter);
+
+        takenSequences.remove(queueId);
+    }
+
+    @Override
+    public boolean contains(@Nonnull QueueIdentifier identifier) {
+        if (logger.isDebugEnabled()) {
+            logger.debug("Validating existence of sequence identifier [{}].", identifier.combinedIdentifier());
+        }
+        return deadLetters.containsKey(identifier);
+    }
+
+    @Override
+    public boolean isFull(@Nonnull QueueIdentifier queueIdentifier) {
+        return maximumNumberOfQueuesReached(queueIdentifier) || maximumQueueSizeReached(queueIdentifier);
+    }
+
+    private boolean maximumNumberOfQueuesReached(QueueIdentifier queueIdentifier) {
+        return !deadLetters.containsKey(queueIdentifier) && deadLetters.keySet().size() >= maxQueues;
+    }
+
+    private boolean maximumQueueSizeReached(QueueIdentifier queueIdentifier) {
+        return deadLetters.containsKey(queueIdentifier) && deadLetters.get(queueIdentifier).size() >= maxQueueSize;
+    }
+
+    @Override
+    public long maxQueues() {
+        return maxQueues;
+    }
+
+    @Override
+    public long maxQueueSize() {
+        return maxQueueSize;
+    }
+
+    @Override
+    public synchronized Optional<DeadLetter<T>> take(@Nonnull String group) {
+        if (deadLetters.isEmpty()) {
+            logger.debug("Queue is empty while taking. Returning an empty optional.");
```

Comment:
I'm not sure this information will be of any value to anyone who isn't intimately familiar with the implementation.
Perhaps we should add a log line before the if stating that we're attempting to take an entry from the DeadLetterQueue. Then, when there are not items in the queue, there is another log line to state that.

Reply:
fSeems fair to me. I'll move this and make it a `trace` if we invoke it for every scenario.